File: Data Structures & Algorithms/three-integer-sum/submission-0.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        ans = []
        n = len(nums)
        
        for i in range(n - 2):
            # 避免重複計算相同的 nums[i]
            if i > 0 and nums[i] == nums[i - 1]:
                continue
            
            target = -nums[i]
            front, rear = i + 1, n - 1

            while front < rear:
                total = nums[front] + nums[rear]
                if total > target:
                    rear -= 1
                elif total < target:
                    front += 1
                else:
                    ans.append([nums[i], nums[front], nums[rear]])

                    # 跳過重複的 front 值
                    while front < rear and nums[front] == nums[front + 1]:
                        front += 1
                    # 跳過重複的 rear 值
                    while front < rear and nums[rear] == nums[rear - 1]:
                        rear -= 1
                    
   
                    front += 1
                    rear -= 1
        
        return ans

File: Data Structures & Algorithms/combination-target-sum-ii/submission-0.cpp
class Solution {
public:
    vector<vector<int>> ans;

    void dfs(vector<int>& candidates, int i, int remain,
             vector<int>& cur) {

        if (remain == 0) {
            ans.push_back(cur);
            return;
        }

        if (i == candidates.size() || candidates[i] > remain) {
            return;
        }

        // 1. 選 candidates[i]
        cur.push_back(candidates[i]);
        dfs(candidates, i + 1, remain - candidates[i], cur);
        cur.pop_back();

        // 2. 不選 candidates[i]
        // 同樣的數字全部跳過，避免產生重複答案
        int next = i + 1;

        while (next < candidates.size() &&
               candidates[next] == candidates[i]) {
            next++;
        }

        dfs(candidates, next, remain, cur);
    }

    vector<vector<int>> combinationSum2(vector<int>& candidates,
                                        int target) {
        sort(candidates.begin(), candidates.end());

        vector<int> cur;
        dfs(candidates, 0, target, cur);

        return ans;
    }
};